File: 337.打家劫舍-iii.go
/*
 * @lc app=leetcode.cn id=337 lang=golang
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rob(root *TreeNode) int {
	max := func(x,y int) int {
		if x > y {
			return x
		}
		return y
	}
	
	var robroot func(root *TreeNode) (int,int) 
	robroot = func(root *TreeNode) (int,int) {
		// 返回值分别是包含root自身和不包含root自身得到的最大偷盗值
		if root.Left == nil && root.Right == nil {
			return root.Val,0
		}
		leftwithroot,leftwithoutroot := 0,0
		rightwithroot,rightwithoutroot := 0,0
		if root.Left != nil {
			leftwithroot,leftwithoutroot = robroot(root.Left)
		}
		if root.Right != nil {
			rightwithroot,rightwithoutroot = robroot(root.Right)
		}
		return root.Val+leftwithoutroot+rightwithoutroot,max(leftwithroot,leftwithoutroot)+max(rightwithroot,rightwithoutroot)
	}

	return max(robroot(root))
}
// @lc code=end

File: 337.打家劫舍-iii.ts
/*
 * @lc app=leetcode.cn id=337 lang=typescript
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rob(root: TreeNode | null): number {
    return Math.max(...dfs(root));
};

function dfs(root: TreeNode | null): [number, number] {
    if (root === null) {
        return [0,0];
    }
    let [lroot, lnoroot] = dfs(root.left);
    let [rroot, rnoroot] = dfs(root.right);
    return [root.val + lnoroot + rnoroot, Math.max(lroot, lnoroot) + Math.max(rroot, rnoroot)]; 
}
// @lc code=end

